File: ios/DailyMemory/Data/Remote/AIAnalysisService.swift
import Foundation

/// AI Analysis Service for extracting entities from text
/// Supports both OpenAI and Claude APIs
actor AIAnalysisService {
    static let shared = AIAnalysisService()

    private let apiClient = APIClient.shared

    private init() {}

    // MARK: - Analysis

    /// Analyze text via backend API (falls back to simulation if offline)
    func analyzeText(_ text: String) async -> Result<AnalysisResult, Error> {
        // Try backend API first
        do {
            let response: AnalysisResponse = try await apiClient.post("ai/analyze", body: ["text": text])
            let personsWithRel = response.persons.map { PersonWithRelationship(from: $0) }
            let extractedTasks = (response.tasks ?? []).map { task in
                ExtractedTask(
                    title: task.title,
                    description: task.description,
                    dueDate: task.dueDate,
                    urgency: task.urgency,
                    importance: task.importance,
                    relatedPerson: task.relatedPerson
                )
            }
            let result = AnalysisResult(
                persons: response.persons.map { $0.name },
                location: response.location,
                date: response.date,
                amount: response.amount,
                tags: response.tags,
                category: response.category,
                mood: response.mood,
                moodScore: response.moodScore,
                summary: response.summary,
                personsWithRelationship: personsWithRel,
                extractedTasks: extractedTasks
            )
            return .success(result)
        } catch {
            print("[AIAnalysis] Backend API error: \(error)")
            // Fallback to simulation on error (offline, rate limited, etc.)
            return .success(simulateAnalysis(text))
        }
    }

    // MARK: - Legacy Response Parsing (kept for reference)

    private func parseAnalysisResponse(_ data: Data) throws -> AnalysisResult {
        // Parse the API response to extract the content
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw AIAnalysisError.invalidResponse
        }

        // Extract content from Claude or OpenAI response
        var content: String?

        // Claude format
        if let contentArray = json["content"] as? [[String: Any]],
           let firstContent = contentArray.first,
           let text = firstContent["text"] as? String {
            content = text
        }

        // OpenAI format
        if content == nil,
           let choices = json["choices"] as? [[String: Any]],
           let firstChoice = choices.first,
           let message = firstChoice["message"] as? [String: Any],
           let text = message["content"] as? String {
            content = text
        }

        guard let contentString = content else {
            throw AIAnalysisError.invalidResponse
        }

        // Extract JSON from content
        guard let jsonStart = contentString.firstIndex(of: "{"),
              let jsonEnd = contentString.lastIndex(of: "}") else {
            throw AIAnalysisError.invalidResponse
        }

        let jsonString = String(contentString[jsonStart...jsonEnd])
        guard let jsonData = jsonString.data(using: .utf8) else {
            throw AIAnalysisError.invalidResponse
        }

        return try JSONDecoder().decode(AnalysisResult.self, from: jsonData)
    }

    // MARK: - Simulation

    /// Simulate AI analysis for testing/demo purposes
    private func simulateAnalysis(_ text: String) -> AnalysisResult {
        let lowerText = text.lowercased()

        // Extract names (simple heuristic: capitalized words)
        let namePattern = try? NSRegularExpression(pattern: "\\b([A-Z][a-z]+(?:\\s+[A-Z][a-z]+)?)\\b")
        let range = NSRange(text.startIndex..., in: text)
        let matches = namePattern?.matches(in: text, range: range) ?? []

        let persons = matches.compactMap { match -> String? in
            guard let range = Range(match.range, in: text) else { return nil }
            let name = String(text[range])
            if name.count > 2 && !Self.commonWords.contains(name.lowercased()) {
                return name
            }
            return nil
        }
        .removingDuplicates()

        // Extract location
        var location: String?
        let locationKeywords = ["at ", "in ", "near ", "to "]
        for keyword in locationKeywords {
            if let range = lowerText.range(of: keyword) {
                let remaining = String(text[range.upperBound...])
                if let endIndex = remaining.firstIndex(where: { ".!?,\n".contains($0) }) {
                    location = String(remaining[..<endIndex]).trimmingCharacters(in: .whitespaces)
                } else {
                    location = remaining.split(separator: " ").prefix(3).joined(separator: " ")
                }
                if let loc = location, !loc.isEmpty { break }
            }
        }

        // Extract amount
        let amountPattern = try? NSRegularExpression(pattern: "\\$([\\d,]+(?:\\.\\d{2})?)")
        let amountMatch = amountPattern?.firstMatch(in: text, range: range)
        var amount: Double?
        if let match = amountMatch, let range = Range(match.range(at: 1), in: text) {
            let amountString = String(text[range]).replacingOccurrences(of: ",", with: "")
            amount = Double(amountString)
        }

        // Determine category
        let category: Category
        if lowerText.contains("wedding") || lowerText.contains("birthday") ||
           lowerText.contains("party") || lowerText.contains("anniversary") {
            category = .event
        } else if lowerText.contains("meeting") || lowerText.contains("lunch") ||
                  lowerText.contains("dinner") || lowerText.contains("met") {
            category = .meeting
        } else if lowerText.contains("pay") || lowerText.contains("owe") ||
                  lowerText.contains("money") || amount != nil {
            category = .financial
        } else if lowerText.contains("promise") || lowerText.contains("will") ||
                  lowerText.contains("remind") || lowerText.contains("todo") {
            category = .promise
        } else {
            category = .general
        }

        // Generate tags
        var tags: [String] = []
        if !persons.isEmpty { tags.append("people") }
        if location != nil { tags.append("location") }
        if amount != nil { tags.append("money") }
        if lowerText.contains("gift") { tags.append("gift") }
        if lowerText.contains("work") || lowerText.contains("office") { tags.append("work") }
        if lowerText.contains("family") { tags.append("family") }

        // Generate summary
        let summary = String(text.prefix(100)) + (text.count > 100 ? "..." : "")

        // Extract simulated tasks from promise-like content
        var simulatedTasks: [ExtractedTask] = []
        let promiseKeywords = ["need to", "have to", "should", "must", "will", "promise", "remind me", "don't forget"]
        for keyword in promiseKeywords {
            if lowerText.contains(keyword) {
                let taskTitle = String(text.prefix(60)).trimmingCharacters(in: .whitespaces)
                simulatedTasks.append(ExtractedTask(
                    title: taskTitle,
                    description: nil,
                    dueDate: nil,
                    urgency: 3,
                    importance: 3,
                    relatedPerson: persons.first
                ))
                break
            }
        }

        return AnalysisResult(
            persons: persons,
            location: location,
            date: nil,
            amount: amount,
            tags: tags,
            category: category.rawValue,
            summary: summary,
            extractedTasks: simulatedTasks
        )
    }

    // MARK: - Constants

    private static let commonWords: Set<String> = [
        "the", "and", "but", "for", "not", "you", "all", "can", "had", "her",
        "was", "one", "our", "out", "day", "get", "has", "him", "his", "how",
        "its", "may", "new", "now", "old", "see", "two", "way", "who", "boy",
        "did", "got", "let", "put", "say", "she", "too", "use", "monday",
        "tuesday", "wednesday", "thursday", "friday", "saturday", "sunday",
        "january", "february", "march", "april", "june", "july", "august",
        "september", "october", "november", "december", "today", "yesterday"
    ]
}

// MARK: - Analysis Result

struct PersonWithRelationship {
    let name: String
    let relationship: Relationship

    init(name: String, relationship: Relationship = .other) {
        self.name = name
        self.relationship = relationship
    }

    init(from extracted: PersonExtracted) {
        self.name = extracted.name
        self.relationship = Relationship(rawValue: extracted.relationship) ?? .other
    }
}

struct AnalysisResult: Codable {
    let persons: [String]
    let location: String?
    let date: String?
    let amount: Double?
    let tags: [String]
    let category: String
    let mood: String?
    let moodScore: Int?
    let summary: String

    // Non-codable: relationship info from AI
    var personsWithRelationship: [PersonWithRelationship] = []
    // AI-extracted tasks/promises
    var extractedTasks: [ExtractedTask] = []

    enum CodingKeys: String, CodingKey {
        case persons, location, date, amount, tags, category, mood, moodScore, summary, extractedTasks
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        persons = try container.decodeIfPresent([String].self, forKey: .persons) ?? []
        location = try container.decodeIfPresent(String.self, forKey: .location)
        date = try container.decodeIfPresent(String.self, forKey: .date)
        amount = try container.decodeIfPresent(Double.self, forKey: .amount)
        tags = try container.decodeIfPresent([String].self, forKey: .tags) ?? []
        category = try container.decodeIfPresent(String.self, forKey: .category) ?? "GENERAL"
        mood = try container.decodeIfPresent(String.self, forKey: .mood)
        moodScore = try container.decodeIfPresent(Int.self, forKey: .moodScore)
        summary = try container.decodeIfPresent(String.self, forKey: .summary) ?? ""
        extractedTasks = try container.decodeIfPresent([ExtractedTask].self, forKey: .extractedTasks) ?? []
        personsWithRelationship = []
    }

    init(
        persons: [String] = [],
        location: String? = nil,
        date: String? = nil,
        amount: Double? = nil,
        tags: [String] = [],
        category: String = "GENERAL",
        mood: String? = nil,
        moodScore: Int? = nil,
        summary: String = "",
        personsWithRelationship: [PersonWithRelationship] = [],
        extractedTasks: [ExtractedTask] = []
    ) {
        self.persons = persons
        self.location = location
        self.date = date
        self.amount = amount
        self.tags = tags
        self.category = category
        self.mood = mood
        self.moodScore = moodScore
        self.summary = summary
        self.personsWithRelationship = personsWithRelationship
        self.extractedTasks = extractedTasks
    }
}

// MARK: - Errors
enum AIAnalysisError: LocalizedError {
    case invalidResponse
    case networkError(Error)

    var errorDescription: String? {
        switch self {
        case .invalidResponse:
            return "Invalid response from AI service"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        }
    }
}

// MARK: - Array Extension
private extension Array where Element: Hashable {
    func removingDuplicates() -> [Element] {
        var seen = Set<Element>()
        return filter { seen.insert($0).inserted }
    }
}

File: android/app/src/main/java/com/dailymemory/data/remote/AIAnalysisService.kt
package com.dailymemory.data.remote

import com.dailymemory.domain.model.Category
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import org.json.JSONArray
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

/**
 * AI Analysis Service for extracting entities from text
 * Supports both OpenAI and Claude APIs
 */
@Singleton
class AIAnalysisService @Inject constructor() {

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()


    // API Configuration - should be stored securely (e.g., BuildConfig, encrypted preferences)
    private var apiKey: String = ""
    private var apiProvider: AIProvider = AIProvider.CLAUDE

    enum class AIProvider {
        OPENAI,
        CLAUDE
    }

    /**
     * Configure the AI service
     */
    fun configure(apiKey: String, provider: AIProvider = AIProvider.CLAUDE) {
        this.apiKey = apiKey
        this.apiProvider = provider
    }

    /**
     * Analyze text and extract entities
     */
    suspend fun analyzeText(text: String): Result<AnalysisResult> = withContext(Dispatchers.IO) {
        if (apiKey.isEmpty()) {
            // Return simulated result when no API key is configured
            return@withContext Result.success(simulateAnalysis(text))
        }

        try {
            val result = when (apiProvider) {
                AIProvider.OPENAI -> analyzeWithOpenAI(text)
                AIProvider.CLAUDE -> analyzeWithClaude(text)
            }
            Result.success(result)
        } catch (e: Exception) {
            // Fallback to simulation on error
            Result.success(simulateAnalysis(text))
        }
    }

    private suspend fun analyzeWithClaude(text: String): AnalysisResult {
        val prompt = buildAnalysisPrompt(text)

        val requestBody = """
            {
                "model": "claude-3-haiku-20240307",
                "max_tokens": 1024,
                "messages": [
                    {
                        "role": "user",
                        "content": "$prompt"
                    }
                ]
            }
        """.trimIndent()

        val request = Request.Builder()
            .url("https://api.anthropic.com/v1/messages")
            .addHeader("Content-Type", "application/json")
            .addHeader("x-api-key", apiKey)
            .addHeader("anthropic-version", "2023-06-01")
            .post(requestBody.toRequestBody("application/json".toMediaType()))
            .build()

        val response = client.newCall(request).execute()
        val responseBody = response.body?.string() ?: throw Exception("Empty response")

        return parseAnalysisResponse(responseBody)
    }

    private suspend fun analyzeWithOpenAI(text: String): AnalysisResult {
        val prompt = buildAnalysisPrompt(text)

        val requestBody = """
            {
                "model": "gpt-3.5-turbo",
                "messages": [
                    {
                        "role": "system",
                        "content": "You are a helpful assistant that extracts structured information from text."
                    },
                    {
                        "role": "user",
                        "content": "$prompt"
                    }
                ],
                "temperature": 0.3
            }
        """.trimIndent()

        val request = Request.Builder()
            .url("https://api.openai.com/v1/chat/completions")
            .addHeader("Content-Type", "application/json")
            .addHeader("Authorization", "Bearer $apiKey")
            .post(requestBody.toRequestBody("application/json".toMediaType()))
            .build()

        val response = client.newCall(request).execute()
        val responseBody = response.body?.string() ?: throw Exception("Empty response")

        return parseAnalysisResponse(responseBody)
    }

    private fun buildAnalysisPrompt(text: String): String {
        return """
            Analyze the following text and extract structured information.
            Return a JSON object with these fields:
            - persons: array of person names mentioned
            - location: the main location mentioned (or null)
            - date: any specific date mentioned in ISO format (or null)
            - amount: any monetary amount mentioned as a number (or null)
            - tags: array of relevant keywords/topics
            - category: one of EVENT, PROMISE, MEETING, FINANCIAL, GENERAL
            - summary: a brief one-sentence summary

            Text: "$text"

            Respond ONLY with valid JSON, no additional text.
        """.trimIndent().replace("\n", "\\n")
    }

    private fun parseAnalysisResponse(responseBody: String): AnalysisResult {
        // Extract JSON content from API response
        val contentRegex = """"content":\s*"([^"]+)"""".toRegex()
        val textRegex = """"text":\s*"([^"]+)"""".toRegex()

        val content = contentRegex.find(responseBody)?.groupValues?.get(1)
            ?: textRegex.find(responseBody)?.groupValues?.get(1)
            ?: responseBody

        // Parse the extracted JSON
        return try {
            val cleanJson = content
                .replace("\\n", "\n")
                .replace("\\\"", "\"")
                .let { str ->
                    val start = str.indexOf("{")
                    val end = str.lastIndexOf("}") + 1
                    if (start >= 0 && end > start) str.substring(start, end) else str
                }

            val jsonObj = JSONObject(cleanJson)

            val persons = mutableListOf<String>()
            jsonObj.optJSONArray("persons")?.let { arr ->
                for (i in 0 until arr.length()) {
                    persons.add(arr.getString(i))
                }
            }

            val tags = mutableListOf<String>()
            jsonObj.optJSONArray("tags")?.let { arr ->
                for (i in 0 until arr.length()) {
                    tags.add(arr.getString(i))
                }
            }

            AnalysisResult(
                persons = persons,
                location = jsonObj.optString("location").takeIf { it.isNotEmpty() && it != "null" },
                date = jsonObj.optString("date").takeIf { it.isNotEmpty() && it != "null" },
                amount = jsonObj.optDouble("amount").takeIf { !it.isNaN() },
                tags = tags,
                category = jsonObj.optString("category", "GENERAL"),
                summary = jsonObj.optString("summary", "")
            )
        } catch (e: Exception) {
            // Return empty result on parse error
            AnalysisResult()
        }
    }

    /**
     * Simulate AI analysis for testing/demo purposes
     */
    private fun simulateAnalysis(text: String): AnalysisResult {
        val lowerText = text.lowercase()

        // Extract names (simple heuristic: capitalized words)
        val namePattern = "\\b([A-Z][a-z]+(?:\\s+[A-Z][a-z]+)?)\\b".toRegex()
        val persons = namePattern.findAll(text)
            .map { it.value }
            .filter { it.length > 2 && !commonWords.contains(it.lowercase()) }
            .distinct()
            .toList()

        // Extract location
        val locationKeywords = listOf("at ", "in ", "near ", "to ")
        var location: String? = null
        for (keyword in locationKeywords) {
            val idx = lowerText.indexOf(keyword)
            if (idx >= 0) {
                val remaining = text.substring(idx + keyword.length)
                val endIdx = remaining.indexOfAny(charArrayOf('.', ',', '!', '?', '\n'))
                location = if (endIdx > 0) remaining.substring(0, endIdx).trim()
                          else remaining.split(" ").take(3).joinToString(" ").trim()
                if (location.isNotEmpty()) break
            }
        }

        // Extract amount
        val amountPattern = "\\$([\\d,]+(?:\\.\\d{2})?)".toRegex()
        val amount = amountPattern.find(text)?.groupValues?.get(1)
            ?.replace(",", "")?.toDoubleOrNull()

        // Determine category
        val category = when {
            lowerText.contains("wedding") || lowerText.contains("birthday") ||
            lowerText.contains("party") || lowerText.contains("anniversary") -> Category.EVENT
            lowerText.contains("meeting") || lowerText.contains("lunch") ||
            lowerText.contains("dinner") || lowerText.contains("met") -> Category.MEETING
            lowerText.contains("pay") || lowerText.contains("owe") ||
            lowerText.contains("money") || amount != null -> Category.FINANCIAL
            lowerText.contains("promise") || lowerText.contains("will") ||
            lowerText.contains("remind") || lowerText.contains("todo") -> Category.PROMISE
            else -> Category.GENERAL
        }

        // Generate tags
        val tags = mutableListOf<String>()
        if (persons.isNotEmpty()) tags.add("people")
        if (location != null) tags.add("location")
        if (amount != null) tags.add("money")
        if (lowerText.contains("gift")) tags.add("gift")
        if (lowerText.contains("work") || lowerText.contains("office")) tags.add("work")
        if (lowerText.contains("family")) tags.add("family")

        // Generate summary
        val summary = text.take(100) + if (text.length > 100) "..." else ""

        return AnalysisResult(
            persons = persons,
            location = location,
            date = null,
            amount = amount,
            tags = tags,
            category = category.name,
            summary = summary
        )
    }

    companion object {
        private val commonWords = setOf(
            "the", "and", "but", "for", "not", "you", "all", "can", "had", "her",
            "was", "one", "our", "out", "day", "get", "has", "him", "his", "how",
            "its", "may", "new", "now", "old", "see", "two", "way", "who", "boy",
            "did", "got", "let", "put", "say", "she", "too", "use", "monday",
            "tuesday", "wednesday", "thursday", "friday", "saturday", "sunday",
            "january", "february", "march", "april", "june", "july", "august",
            "september", "october", "november", "december", "today", "yesterday"
        )
    }
}

data class AnalysisResult(
    val persons: List<String> = emptyList(),
    val location: String? = null,
    val date: String? = null,
    val amount: Double? = null,
    val tags: List<String> = emptyList(),
    val category: String = "GENERAL",
    val summary: String = ""
)
